packages/sitecore-jss-react/src/enhancers/withSitecorePersonalizationContext.tsx

```diff
@@ -0,0 +1,58 @@
+import React from 'react';
+import {
+  isServer,
+  LayoutPersonalizationService,
+  LayoutServiceData,
+  SitecorePersonalizationContextState,
+} from '@sitecore-jss/sitecore-jss';
+
+export const SitecorePersonalizationReactContext = React.createContext<
+  SitecorePersonalizationContextState | undefined
+>(undefined);
+
+export interface SitecorePersonalizationContextProps {
+  layoutData: LayoutServiceData | null;
+  isPreview: boolean;
+  isPersonalizationSuppressed?: boolean;
+  tracked: boolean;
+}
+
+/**
+ * @param {React.ComponentClass<T> | React.SFC<T>} Component
+ * @param {LayoutPersonalizationService} layoutPersonalizationService
+ */
+export function withSitecorePersonalizationContext<T extends SitecorePersonalizationContextProps>(
+  Component: React.ComponentClass<T> | React.SFC<T>,
+  layoutPersonalizationService: LayoutPersonalizationService
+) {
+  return function WithSitecorePersonalizationContext(props: T) {
+    const route = props.layoutData?.sitecore.route;
+
+    let personalizationContext: SitecorePersonalizationContextState | undefined = undefined;
+    const newProps = { ...props };
+    if (
+      !props.isPersonalizationSuppressed &&
+      !props.isPreview &&
+      route &&
+      route.layoutId !== 'available-in-connected-mode' &&
+      !isServer()
+    ) {
+      personalizationContext = layoutPersonalizationService.startPersonalization(
+        {
+          language: props.layoutData?.sitecore.context.language as string,
+          routePath: props.layoutData?.sitecore.context.itemPath as string,
+        },
+        route
+      );
+      if (personalizationContext) {
+        newProps.tracked = personalizationContext.isTracked;
+      }
+    }
+
+    return (
+      <SitecorePersonalizationReactContext.Provider value={personalizationContext}>
```

Comment:
- So in your case, it means, that we will use new context on each call?

Yes, this is exactly what we want to achieve. This HOC starts personalization of the whole page and **provides** all nested personalizable components with an instance of personalization context for they can use it to check personalization status and **consume** personalization results.

Reply:
@maximvsidorenko Probably, I still didn't get idea, if you have a single Context, you can use our approach:
https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-react/src/components/SitecoreContext.tsx#L23
https://github.com/Sitecore/jss/blob/dev/packages/sitecore-jss-react/src/components/SitecoreContext.tsx#L64-L68
What is the difference between our approach and that? 
`withSitecorePersonalizationContext` means that it will provide you data which already exists in some context, but for this purpose you create a new context each time. If follow your approach, you don't need even a React.context here.
You can have a separate service to call that data and return it, instead of having separate storage for each component. Need to discuss it